utils/yoke/src/yoke.rs

```diff
@@ -480,50 +472,200 @@ where
     }
 }
 
-impl<Y: for<'a> Yokeable<'a>, T: ?Sized> Clone for Yoke<Y, Option<Rc<T>>>
-where
-    for<'a> <Y as Yokeable<'a>>::Output: Clone,
-{
-    fn clone(&self) -> Self {
+impl<Y: for<'a> Yokeable<'a>, C: CloneableCart> Yoke<Y, C> {
+    /// Allows one to "project" a yoke to perform a transformation on the data, potentially
+    /// looking at a subfield, and producing a new yoke. This will clone the cart, and the provided
+    /// transformation is only allowed to use data known to be borrowed from the cart.
+    ///
+    /// This takes an additional `PhantomData<&()>` parameter as a workaround to the issue
+    /// described in [#86702](https://github.com/rust-lang/rust/issues/86702). This parameter
+    /// should just be ignored in the function.
+    /// 
+    /// Furthermore,
+    /// [compiler bug #84937](https://github.com/rust-lang/rust/issues/84937) prevents
+    /// this from taking a capturing closure, however [`Yoke::project_with_capture()`]
+    /// can be used for the same use cases.
+    ///
+    ///
+    /// This can be used, for example, to transform data from one format to another:
+    ///
+    /// ```rust
+    /// # use std::rc::Rc;
+    /// # use yoke::Yoke;
+    /// #
+    /// fn slice(y: Yoke<&'static str, Rc<[u8]>>) -> Yoke<&'static [u8], Rc<[u8]>> {
+    ///    y.project(move |yk, _| yk.as_bytes())
+    /// }
+    ///
+    /// ```
+    ///
+    /// This can also be used to create a yoke for a subfield
+    ///
+    /// ```rust
+    /// # use std::borrow::Cow;
+    /// # use yoke::{Yoke, Yokeable};
+    /// # use std::mem;
+    /// # use std::rc::Rc;
+    /// #
+    /// // also safely implements Yokeable<'a>
+    /// struct Bar<'a> {
+    ///     string_1: &'a str,
+    ///     string_2: &'a str,
+    /// }
+    ///
+    /// fn project_string_1(bar: &Yoke<Bar<'static>, Rc<[u8]>>) -> Yoke<&'static str, Rc<[u8]>> {
+    ///     bar.project(|bar, _| bar.string_1)   
+    /// }
+    ///
+    /// #
+    /// # unsafe impl<'a> Yokeable<'a> for Bar<'static> {
+    /// #     type Output = Bar<'a>;
+    /// #     fn transform(&'a self) -> &'a Bar<'a> {
+    /// #         self
+    /// #     }
+    /// #
+    /// #     unsafe fn make(from: Bar<'a>) -> Self {
+    /// #         let ret = mem::transmute_copy(&from);
+    /// #         mem::forget(from);
+    /// #         ret
+    /// #     }
+    /// #
+    /// #     fn transform_mut<F>(&'a mut self, f: F)
+    /// #     where
+    /// #         F: 'static + FnOnce(&'a mut Self::Output),
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    //
+    // Safety docs can be found below on `__project_safety_docs()`
+    pub fn project<'this, P>(
+        &'this self,
```

Comment:
Suggestion: in most cases I think we can move `self` and avoid the `self.cart.clone()`; if the caller doesn't want to move `self`, they can `Yoke::clone()` first.  That would align more with my mental model on how `.project()` should be used.  As written, `.project()` is really more like `.clone_project()`.

Reply:
That works for me!

Reply:
Oh, actually, it's trickier, because of the lifetimes: this is a bit easier to reason about because we're borrowing, not dealing with fully owned inputs. Let me think about it

Reply:
It might be worth having both.  We should always have `.project(self)`, and if `C: CloneableCart`, then we can add `.project_clone(&self)`.

Reply:
Okay, I was able to make it moving for `self`, however the closure still is given a borrowed yokeable. There is a way to remove that requirement but it's a bunch more work and requires an additional method on Yoke. Is this fine for now or do you want this to operate on owned yokeables too?

Reply:
I think I'd prefer to try that in a followup so as not to complicate the safety reasoning so far.